### PR TITLE
Add default task weights 

### PR DESCRIPTION
**Pull request by KnathanM**
## Description
In #779 we moved task weights to be an attribute of LossFunction. This means it is also an attribute of Metric even though it isn't used. We will address this more thoroughly in v2.1, but requiring task weights when creating a metric will be confusing for users. I thought about it some more and feel task weight should also not be required for when making a loss function if all task weights should be one.

Before I didn't put default task weights of ones in the init for LossFunction because in the LossFunction we don't know how many tasks there will be. Compare to predictors.py that uses `torch.ones(n_tasks) to make a default task weights. But then I realized if we want all ones, we don't have to have the length right because it can be broadcasted. 

This PR allows for simple creation of metrics and loss functions if the task weights should be ones via `MSEMetric()`/`MSELoss()`. I chose to not include task weights in the repr string if it is a tensor of a single 1 as `None` is the real default and if `task_weights == torch.ones(1).view(1,-1)` that is just a way to get the calculations to work.

### CODE REVIEW

**Review by shihchengli — APPROVED** [review]

File: chemprop/nn/loss.py [hunk 1]
```diff
@@ -27,15 +27,19 @@
 
 
 class LossFunction(nn.Module):
-    def __init__(self, task_weights: Tensor):
+    def __init__(self, task_weights: Tensor | None = None):
```
Comment (davidegraff) [inline chemprop/nn/loss.py:30]: does this need to be `Tensor | None`? what about `task_weights: ArrayLike = 1.0`? This corresponds to a default of equal-weight tasks as scalars are fully broadcasted.

File: chemprop/nn/metrics.py [hunk 2]
```diff
@@ -44,18 +44,6 @@
 
 
 class Metric(LossFunction):
```
Comment (davidegraff) [inline chemprop/nn/metrics.py:46]: So am I understanding correctly that a user can supply `task_weights` to a `Metric` but that they won't be used? If so, why are you deleting the note in the docstring?

File: chemprop/nn/loss.py [hunk 3]
```diff
@@ -81,6 +85,8 @@ def _calc_unreduced_loss(self, preds, targets, mask, weights, lt_mask, gt_mask)
         """Calculate a tensor of shape `b x t` containing the unreduced loss values."""
 
     def extra_repr(self) -> str:
+        if self.task_weights == torch.ones(1).view(1, -1):
+            return ""
         return f"task_weights={self.task_weights.tolist()}"
```
Comment (davidegraff) [inline chemprop/nn/loss.py:90]:
```suggestion
        return f"task_weights={self.task_weights.tolist()}"
```

It's fine to display defaults. See pytorch defaults:
```python
>>> nn.Dropout()
Dropout(p=0.5, inplace=False)
```

File: chemprop/nn/loss.py (hunk 1 above)
Comment (KnathanM) [reply chemprop/nn/loss.py:30]: I thought about it more and agree 1.0 is a better default than None.

File: chemprop/nn/metrics.py (hunk 2 above)
Comment (KnathanM) [reply chemprop/nn/metrics.py:46]: Good point, I'll add back a note that task weights aren't used to the Metric docstring.

File: chemprop/nn/loss.py [hunk 4]
```diff
@@ -27,12 +28,12 @@
...
         """
         Parameters
         ----------
-        task_weights : Tensor
-            a tensor of shape `t` or `1 x t` containing the per-task weight.
+        task_weights :  ArrayLike = 1.0
+            the per-task weights of shape `t` or `1 x t`. Defaults to all tasks having a weight of 1.
         """
         super().__init__()
         task_weights = torch.as_tensor(task_weights, dtype=torch.float32).view(1, -1)
```
Comment (davidegraff) [inline chemprop/nn/loss.py:39]:
```suggestion
        task_weights = torch.as_tensor(task_weights, dtype=torch.float).view(1, -1)
```

unless you have good reason to specify precision, prefer high-level aliases (same goes for numpy, c.f. `np.single` vs. `np.float32`)

File: chemprop/nn/metrics.py [hunk 5]
```diff
@@ -45,15 +45,12 @@
...
     """
-    __init__(self, task_weights: Tensor):
+    __init__(self, task_weights: ArrayLike = 1.0):
```
Comment (davidegraff) [inline chemprop/nn/metrics.py:48]: why is the method signature inside the docstring?

File: chemprop/nn/loss.py [hunk 6]
```diff
@@ -81,7 +82,7 @@ def _calc_unreduced_loss(self, preds, targets, mask, weights, lt_mask, gt_mask)
         """Calculate a tensor of shape `b x t` containing the unreduced loss values."""
 
     def extra_repr(self) -> str:
-        return f"task_weights={self.task_weights.tolist()}"
+        return f"task_weights={self.task_weights}"
```
Comment (davidegraff) [inline chemprop/nn/loss.py:85]:
```suggestion
        return f"task_weights={self.task_weights.tolist()}"
```
IMO a list is a clean `repr` than the raw tensor

File: chemprop/nn/metrics.py [hunk 7]
```diff
@@ -45,15 +45,12 @@
...
 
     Parameters
     ----------
-    task_weights : Tensor
-        a tensor of shape `t` or `1 x t` containing the per-task weight.
-    .. note::
-        This class inherits from :class:`LossFunction`, which requires task_weights. These
-        task_weights are not used in Metric and can be set to any Tensor, e.g. tensor([1.])
+    task_weights :  ArrayLike = 1.0
+        Ignored. Inherited from :class:`LossFunction`, but not used in :class:`Metric`.
```
Comment (davidegraff) [inline chemprop/nn/metrics.py:53]:
```suggestion
        .. important::
            Ignored. Maintained for compatibility with :class:`~chemprop.nn.loss.LossFunction`
```

File: chemprop/nn/metrics.py (hunk 5 above)
Comment (KnathanM) [reply chemprop/nn/metrics.py:48]: I thought I had to specify which method the Parameters shown were for, which was difficult given the method is defined in LossFunction. I now see that a class docstring can have Parameters specified, like in `MPNN`. will remove this now.

**Review by davidegraff — APPROVED** [review]
